src/pyramid_retry/__init__.py

```diff
@@ -81,22 +81,29 @@ def retry_policy(environ, router):
         # make the original request
         request_ctx = router.request_context(environ)
         request = request_ctx.begin()
-
-        if activate_hook:
-            retry_attempts = activate_hook(request)
-            if retry_attempts is None:
-                retry_attempts = attempts
+        try:
+            if activate_hook:
+                retry_attempts = activate_hook(request)
+                if retry_attempts is None:
+                    retry_attempts = attempts
+                else:
+                    assert retry_attempts > 0
             else:
-                assert retry_attempts > 0
-        else:
-            retry_attempts = attempts
-
-        # if we are supporting multiple attempts then we must make
-        # make the body seekable in order to re-use it across multiple
-        # attempts. make_body_seekable will copy wsgi.input if
-        # necessary, otherwise it will rewind the copy to position zero
-        if retry_attempts != 1:
-            request.make_body_seekable()
+                retry_attempts = attempts
+
+            # if we are supporting multiple attempts then we must make
+            # make the body seekable in order to re-use it across multiple
+            # attempts. make_body_seekable will copy wsgi.input if
+            # necessary, otherwise it will rewind the copy to position zero
+            if retry_attempts != 1:
+                request.make_body_seekable()
+
+        # Catch make_body_seekable (e.g. 408 RequestTimeout)
+        # and activate_hook exceptions and clean up. Use BaseException
+        # here to catch e.g. GeneratorExit.
+        except BaseException:
```

Comment:
This is the only part of the change that I am slightly worried about. When does a `GeneratorExit` happen? `wsgi.input` should be a file like object according to the WSGI PEP, and it shouldn't be causing a `GeneratorExit`. Is this to safeguard against the `activate_hook` in some way?

Reply:
I don't think it's wrong to catch BaseException here but I'm also curious why GeneratorExit is singled-out as an issue

Reply:
Anyway, this PR is good, and I'm happy with it - just want to get clarity on your comment here and maybe improve it before I merge it.

Reply:
Initially I was only catching `Exception` but changed it to `BaseException` on suggestion from @lrowe. You can see reasoning here: https://github.com/ENCODE-DCC/pyramid_retry/pull/1

`GeneratorExit` inherits directly from `BaseException`, so won't be caught by `Exception`. It seems plausible that user code (for example `activate_hook`) could generate such an exception, and unlike something like `SystemExit` (which also inherits directly from `BaseException`) won't necessarily kill the process (and clean the stack for us).

Reply:
> Anyway, this PR is good, and I'm happy with it - just want to get clarity on your comment here and maybe improve it before I merge it.

Yes, by the way I'm happy to remove/reword any of these comments in code/tests. Only added to clarify my reasoning.

Reply:
Sure I mean it's correct to pop the stack in every situation, so I'm not arguing against BaseException. I think maybe just remove the last sentence.